crates/oxc_query/src/adapter.rs

```diff
@@ -0,0 +1,436 @@
+use std::{
+    rc::Rc,
+    sync::{Arc, OnceLock},
+};
+
+use oxc_semantic::Semantic;
+use trustfall::{
+    provider::{
+        resolve_coercion_using_schema, resolve_property_with, ContextIterator,
+        ContextOutcomeIterator, EdgeParameters, ResolveEdgeInfo, ResolveInfo, Typename,
+        VertexIterator,
+    },
+    FieldValue, Schema,
+};
+
+use super::vertex::Vertex;
+
+static SCHEMA: OnceLock<Schema> = OnceLock::new();
+
+pub struct Adapter<'a> {
+    pub semantic: Rc<Semantic<'a>>,
+    pub path_components: Vec<Option<String>>,
+}
+
+const SCHEMA_TEXT: &str = include_str!("./schema.graphql");
+
+#[allow(clippy::missing_panics_doc)]
+pub fn schema() -> &'static Schema {
+    SCHEMA.get_or_init(|| Schema::parse(SCHEMA_TEXT).expect("not a valid schema"))
+}
+
+impl<'a, 'b: 'a> trustfall::provider::Adapter<'a> for &'a Adapter<'b> {
+    type Vertex = Vertex<'b>;
+
+    fn resolve_starting_vertices(
+        &self,
+        edge_name: &Arc<str>,
+        _parameters: &EdgeParameters,
+        resolve_info: &ResolveInfo,
+    ) -> VertexIterator<'a, Self::Vertex> {
+        match edge_name.as_ref() {
+            "File" => super::entrypoints::file(resolve_info),
+            _ => {
+                unreachable!(
+                    "attempted to resolve starting vertices for unexpected edge name: {edge_name}"
+                )
+            }
+        }
+    }
+
+    #[allow(clippy::too_many_lines)]
+    fn resolve_property(
+        &self,
+        contexts: ContextIterator<'a, Self::Vertex>,
+        type_name: &Arc<str>,
+        property_name: &Arc<str>,
+        resolve_info: &ResolveInfo,
+    ) -> ContextOutcomeIterator<'a, Self::Vertex, FieldValue> {
+        if property_name.as_ref() == "__typename" {
+            return resolve_property_with(contexts, |v| v.typename().into());
+        }
+        match type_name.as_ref() {
+            "AssignmentType" => super::properties::resolve_assignment_type_property(
+                contexts,
+                property_name.as_ref(),
+                resolve_info,
+            ),
+            "ClassAST" | "Class" => super::properties::resolve_class_property(
```

Comment:
Should this be seperated out to be similar to the rest of the *AST variants, or is it better to reduce code duplication and just point all *AST variants and the non-ast variants to the same methods? (we already just call into the non-ast variant methods, but is it worth just making them both point to the same resolver?)

Reply:
Will they ever diverge such that one of the types has properties the other does not?

Reply:
No.

Reply:
I decided for the purpose of properties, we can group on */*AST together, but for now not for edges since there are most edges on *AST's